File: crates/tui/src/app/dialogs/agents.rs
use super::super::App;
use crate::app::ops::AppOp;
use crate::render::{AgentSnapshot, SharedSnapshots};
use crate::utils::format_duration;
use crossterm::event::{KeyCode, KeyModifiers};
use engine::registry::{AgentStatus, RegistryEntry};
use std::cell::RefCell;
use std::rc::Rc;
use ui::{Callback, CallbackResult, KeyBind, Payload, WinEvent};

struct AgentsListState {
    my_pid: u32,
    snapshots: SharedSnapshots,
    list_buf: ui::BufId,
    agents: Vec<RegistryEntry>,
}

struct AgentsDetailState {
    my_pid: u32,
    agent_id: String,
    snapshots: SharedSnapshots,
    title_buf: ui::BufId,
    detail_buf: ui::BufId,
    parent_selected: usize,
}

pub(in crate::app) fn open(app: &mut App) {
    open_list(app, 0);
}

pub(in crate::app) fn open_list(app: &mut App, initial_selected: usize) {
    use crate::keymap::hints;

    let my_pid = std::process::id();
    let snapshots = app.agent_snapshots.clone();
    let agents = engine::registry::children_of(my_pid);

    let title_buf = app.ui.buf_create(ui::buffer::BufCreateOpts {
        buftype: ui::buffer::BufType::Scratch,
        modifiable: false,
    });
    if let Some(buf) = app.ui.buf_mut(title_buf) {
        buf.set_all_lines(vec!["agents".into(), String::new()]);
        buf.add_highlight(0, 0, 6, ui::buffer::SpanStyle::dim());
    }

    let list_buf = app.ui.buf_create(ui::buffer::BufCreateOpts {
        buftype: ui::buffer::BufType::Scratch,
        modifiable: false,
    });
    refresh_list(&mut app.ui, list_buf, &agents, &snapshots);

    let hint_text = hints::join(&["enter: view", hints::KILL_PROC, hints::CLOSE]);
    let dialog_config = app.builtin_dialog_config(
        Some(hint_text),
        vec![(KeyCode::Char('q'), KeyModifiers::NONE)],
    );

    let Some(win_id) = app.ui.dialog_open(
        ui::FloatConfig {
            title: None,
            border: ui::Border::None,
            placement: ui::Placement::dock_bottom_full_width(ui::Constraint::Pct(60)),
            ..Default::default()
        },
        dialog_config,
        vec![
            ui::PanelSpec::content(title_buf, ui::PanelHeight::Fixed(2)).focusable(false),
            ui::PanelSpec::list(list_buf, ui::PanelHeight::Fill),
        ],
    ) else {
        return;
    };

    if initial_selected > 0 {
        if let Some(dialog) = app.ui.dialog_mut(win_id) {
            dialog.set_selected_index(initial_selected);
        }
    }

    let state = Rc::new(RefCell::new(AgentsListState {
        my_pid,
        snapshots,
        list_buf,
        agents,
    }));

    let ops = app.lua.ops_handle();

    // Backspace: kill the selected subagent if it belongs to our tree.
    let state_bs = state.clone();
    app.ui.win_set_keymap(
        win_id,
        KeyBind::plain(KeyCode::Backspace),
        Callback::Rust(Box::new(move |ctx| {
            let idx = ctx.ui.dialog_mut(ctx.win).and_then(|d| d.selected_index());
            let Some(idx) = idx else {
                return CallbackResult::Consumed;
            };
            let mut s = state_bs.borrow_mut();
            if let Some(agent) = s.agents.get(idx) {
                let pid = agent.pid;
                if engine::registry::is_in_tree(pid, s.my_pid) {
                    engine::registry::kill_agent(pid);
                    s.agents = engine::registry::children_of(s.my_pid);
                    let s_ref: &AgentsListState = &s;
                    refresh_list(ctx.ui, s_ref.list_buf, &s_ref.agents, &s_ref.snapshots);
                }
            }
            CallbackResult::Consumed
        })),
    );

    // Submit (Enter): swap list for detail view of the selected agent.
    let state_submit = state.clone();
    let ops_submit = ops.clone();
    app.ui.win_on_event(
        win_id,
        WinEvent::Submit,
        Callback::Rust(Box::new(move |ctx| {
            if let Payload::Selection { index } = ctx.payload {
                let s = state_submit.borrow();
                if let Some(entry) = s.agents.get(index) {
                    ops_submit.push(AppOp::OpenAgentsDetail {
                        agent_id: entry.agent_id.clone(),
                        parent_selected: index,
                    });
                }
            }
            ops_submit.push(AppOp::CloseFloat(ctx.win));
            CallbackResult::Consumed
        })),
    );

    // Dismiss: refresh agent counts + close.
    let ops_dismiss = ops.clone();
    app.ui.win_on_event(
        win_id,
        WinEvent::Dismiss,
        Callback::Rust(Box::new(move |ctx| {
            ops_dismiss.push(AppOp::RefreshAgentCounts);
            ops_dismiss.push(AppOp::CloseFloat(ctx.win));
            CallbackResult::Consumed
        })),
    );

    // Tick: re-read registry and refresh the buffer when anything changed.
    let state_tick = state;
    app.ui.win_on_event(
        win_id,
        WinEvent::Tick,
        Callback::Rust(Box::new(move |ctx| {
            let mut s = state_tick.borrow_mut();
            let fresh = engine::registry::children_of(s.my_pid);
            let changed = fresh.len() != s.agents.len()
                || fresh.iter().zip(s.agents.iter()).any(|(a, b)| {
                    a.agent_id != b.agent_id
                        || a.status != b.status
                        || a.task_slug != b.task_slug
                        || a.pid != b.pid
                });
            if changed {
                s.agents = fresh;
                let s_ref: &AgentsListState = &s;
                refresh_list(ctx.ui, s_ref.list_buf, &s_ref.agents, &s_ref.snapshots);
            }
            CallbackResult::Consumed
        })),
    );
    let _ = ops;
}

pub(in crate::app) fn open_detail(app: &mut App, agent_id: String, parent_selected: usize) {
    use crate::keymap::hints;

    let my_pid = std::process::id();
    let snapshots = app.agent_snapshots.clone();

    let title_buf = app.ui.buf_create(ui::buffer::BufCreateOpts {
        buftype: ui::buffer::BufType::Scratch,
        modifiable: false,
    });
    refresh_detail_title(&mut app.ui, title_buf, &agent_id, my_pid, &snapshots);

    let detail_buf = app.ui.buf_create(ui::buffer::BufCreateOpts {
        buftype: ui::buffer::BufType::Scratch,
        modifiable: false,
    });
    refresh_detail_body(&mut app.ui, detail_buf, &agent_id, &snapshots);

    let hint_text = hints::join(&[hints::BACK, hints::scroll(app.input.vim_enabled())]);
    let dialog_config = app.builtin_dialog_config(Some(hint_text), vec![]);

    let Some(win_id) = app.ui.dialog_open(
        ui::FloatConfig {
            title: None,
            border: ui::Border::None,
            placement: ui::Placement::dock_bottom_full_width(ui::Constraint::Pct(60)),
            ..Default::default()
        },
        dialog_config,
        vec![
            ui::PanelSpec::content(title_buf, ui::PanelHeight::Fixed(2)).focusable(false),
            ui::PanelSpec::content(detail_buf, ui::PanelHeight::Fill)
                .with_pad_left(2)
                .focusable(true),
        ],
    ) else {
        return;
    };

    let state = Rc::new(RefCell::new(AgentsDetailState {
        my_pid,
        agent_id,
        snapshots,
        title_buf,
        detail_buf,
        parent_selected,
    }));

    let ops = app.lua.ops_handle();

    // Dismiss: back-nav to the list view.
    let state_dismiss = state.clone();
    let ops_dismiss = ops.clone();
    app.ui.win_on_event(
        win_id,
        WinEvent::Dismiss,
        Callback::Rust(Box::new(move |ctx| {
            let initial = state_dismiss.borrow().parent_selected;
            ops_dismiss.push(AppOp::CloseFloat(ctx.win));
            ops_dismiss.push(AppOp::OpenAgentsList {
                initial_selected: initial,
            });
            CallbackResult::Consumed
        })),
    );

    // Tick: refresh title + body.
    app.ui.win_on_event(
        win_id,
        WinEvent::Tick,
        Callback::Rust(Box::new(move |ctx| {
            let s = state.borrow();
            refresh_detail_title(ctx.ui, s.title_buf, &s.agent_id, s.my_pid, &s.snapshots);
            refresh_detail_body(ctx.ui, s.detail_buf, &s.agent_id, &s.snapshots);
            CallbackResult::Consumed
        })),
    );
    let _ = ops;
}

fn find_snapshot(snapshots: &SharedSnapshots, agent_id: &str) -> Option<AgentSnapshot> {
    let snaps = snapshots.lock().unwrap();
    snaps.iter().find(|s| s.agent_id == agent_id).cloned()
}

fn refresh_list(
    ui: &mut ui::Ui,
    list_buf: ui::BufId,
    agents: &[RegistryEntry],
    snapshots: &SharedSnapshots,
) {
    let name_w = agents.iter().map(|a| a.agent_id.len()).max().unwrap_or(0);

    if agents.is_empty() {
        let Some(buf) = ui.buf_mut(list_buf) else {
            return;
        };
        buf.set_all_lines(vec!["  No subagents running".into()]);
        let len = buf
            .get_line(0)
            .map(|l| l.chars().count() as u16)
            .unwrap_or(0);
        buf.add_highlight(0, 0, len, ui::buffer::SpanStyle::dim());
        return;
    }

    let mut lines: Vec<String> = Vec::with_capacity(agents.len());
    let mut status_spans: Vec<(u16, u16)> = Vec::with_capacity(agents.len());
    for agent in agents {
        let status_str = match agent.status {
            AgentStatus::Working => "working",
            AgentStatus::Idle => "idle   ",
        };
        let name = format!("{:<name_w$}", agent.agent_id);
        let mut line = format!("  {name}  {status_str}");
        let status_start = (2 + name_w + 2) as u16;
        let status_end = status_start + status_str.len() as u16;
        status_spans.push((status_start, status_end));
        if let Some(slug) = &agent.task_slug {
            line.push_str("  ");
            line.push_str(slug);
        }
        if let Some(snap) = find_snapshot(snapshots, &agent.agent_id) {
            if let Some(tokens) = snap.context_tokens {
                line.push_str(&format!("  {}", crate::render::format_tokens(tokens)));
            }
            if snap.cost_usd > 0.0 {
                line.push_str(&format!("  {}", crate::metrics::format_cost(snap.cost_usd)));
            }
        }
        lines.push(line);
    }

    let Some(buf) = ui.buf_mut(list_buf) else {
        return;
    };
    buf.set_all_lines(lines);
    for (i, &(start, end)) in status_spans.iter().enumerate() {
        buf.add_highlight(i, start, end, ui::buffer::SpanStyle::dim());
    }
}

fn refresh_detail_title(
    ui: &mut ui::Ui,
    title_buf: ui::BufId,
    agent_id: &str,
    my_pid: u32,
    snapshots: &SharedSnapshots,
) {
    let Some(buf) = ui.buf_mut(title_buf) else {
        return;
    };
    let entry = engine::registry::children_of(my_pid)
        .into_iter()
        .find(|e| e.agent_id == agent_id);
    let mut line = format!(" {agent_id}");
    let id_end = line.chars().count() as u16;
    if let Some(ref e) = entry {
        if matches!(e.status, AgentStatus::Idle) {
            line.push_str(" \u{2713}");
        }
        if let Some(ref slug) = e.task_slug {
            line.push_str(&format!(" \u{00b7} {slug}"));
        }
    }
    let snap = find_snapshot(snapshots, agent_id);
    if let Some(ref s) = snap {
        if let Some(tokens) = s.context_tokens {
            line.push_str(&format!("  {}", crate::render::format_tokens(tokens)));
        }
        if s.cost_usd > 0.0 {
            line.push_str(&format!("  {}", crate::metrics::format_cost(s.cost_usd)));
        }
    }
    buf.set_all_lines(vec![line, String::new()]);
    buf.add_highlight(
        0,
        1,
        id_end,
        ui::buffer::SpanStyle {
            fg: Some(crate::theme::AGENT),
            bold: true,
            ..Default::default()
        },
    );
}

fn refresh_detail_body(
    ui: &mut ui::Ui,
    detail_buf: ui::BufId,
    agent_id: &str,
    snapshots: &SharedSnapshots,
) {
    let snap = find_snapshot(snapshots, agent_id);
    let mut lines: Vec<String> = Vec::new();
    let mut dim_lines: Vec<usize> = Vec::new();

    let Some(snap) = snap else {
        let Some(buf) = ui.buf_mut(detail_buf) else {
            return;
        };
        buf.set_all_lines(vec!["(agent not tracked)".into()]);
        return;
    };

    dim_lines.push(lines.len());
    lines.push("Prompt:".into());
    for raw_line in snap.prompt.lines() {
        lines.push(format!(" {raw_line}"));
    }
    lines.push(String::new());
    if snap.tool_calls.is_empty() {
        lines.push("(no tool calls yet)".into());
    } else {
        for entry in &snap.tool_calls {
            let elapsed = entry
                .elapsed
                .filter(|d| d.as_secs_f64() >= 0.1)
                .map(|d| format!("  {}", format_duration(d.as_secs())))
                .unwrap_or_default();
            lines.push(format!("{} {}{elapsed}", entry.tool_name, entry.summary));
        }
    }

    let Some(buf) = ui.buf_mut(detail_buf) else {
        return;
    };
    buf.set_all_lines(lines.clone());
    for i in dim_lines {
        if let Some(line) = lines.get(i) {
            let len = line.chars().count() as u16;
            buf.add_highlight(i, 0, len, ui::buffer::SpanStyle::dim());
        }
    }
}

File: runtime/lua/smelt/plugins/agents.lua
-- Built-in /agents command.
--
-- Two views stitched together with a single `smelt.spawn`:
--   1. List view  — one row per subagent with live status, tokens, cost.
--                   Backspace kills the selected agent; Enter opens its
--                   detail view.
--   2. Detail view — prompt + tool-call log for one agent, live-updated
--                    via `on_tick` + `smelt.agent.snapshots`.
-- Dismissing the detail view navigates back to the list.

local function format_duration(secs)
  if secs < 60 then
    return string.format("%ds", secs)
  elseif secs < 3600 then
    return string.format("%dm %ds", secs // 60, secs % 60)
  else
    local h = secs // 3600
    local rest = secs % 3600
    return string.format("%dh %dm %ds", h, rest // 60, rest % 60)
  end
end

local function format_tokens(n)
  if n >= 1000000 then
    return string.format("%.1fm", n / 1000000)
  elseif n >= 1000 then
    return string.format("%.1fk", n / 1000)
  else
    return string.format("%d", n)
  end
end

local function format_cost(usd)
  if usd < 0.01 then
    return string.format("$%.4f", usd)
  elseif usd < 1.0 then
    return string.format("$%.3f", usd)
  else
    return string.format("$%.2f", usd)
  end
end

local function find_snapshot(snapshots, agent_id)
  for _, s in ipairs(snapshots) do
    if s.agent_id == agent_id then return s end
  end
  return nil
end

-- Render the agent registry into `list_buf`.
local function refresh_list(list_buf, agents, snapshots)
  if #agents == 0 then
    smelt.api.buf.set_lines(list_buf, { "  No subagents running" })
    smelt.api.buf.add_dim(list_buf, 1, 0, #"  No subagents running")
    return
  end

  local name_w = 0
  for _, a in ipairs(agents) do
    if #a.agent_id > name_w then name_w = #a.agent_id end
  end

  local lines = {}
  local status_spans = {}
  for _, a in ipairs(agents) do
    local status_str = (a.status == "working") and "working" or "idle   "
    local name = a.agent_id .. string.rep(" ", name_w - #a.agent_id)
    local line = string.format("  %s  %s", name, status_str)
    local status_start = 2 + name_w + 2
    local status_end = status_start + #status_str
    table.insert(status_spans, { status_start, status_end })
    if a.task_slug and a.task_slug ~= "" then
      line = line .. "  " .. a.task_slug
    end
    local snap = find_snapshot(snapshots, a.agent_id)
    if snap then
      if snap.context_tokens and snap.context_tokens > 0 then
        line = line .. "  " .. format_tokens(snap.context_tokens)
      end
      if snap.cost_usd and snap.cost_usd > 0 then
        line = line .. "  " .. format_cost(snap.cost_usd)
      end
    end
    table.insert(lines, line)
  end
  smelt.api.buf.set_lines(list_buf, lines)
  for i, span in ipairs(status_spans) do
    smelt.api.buf.add_dim(list_buf, i, span[1], span[2])
  end
end

-- Compare two agent lists for the fields that drive the rendered row.
-- Tick callbacks refresh the buffer only when something visible changed.
local function agents_changed(a, b)
  if #a ~= #b then return true end
  for i = 1, #a do
    local x, y = a[i], b[i]
    if x.agent_id ~= y.agent_id
        or x.status ~= y.status
        or x.task_slug ~= y.task_slug
        or x.pid ~= y.pid then
      return true
    end
  end
  return false
end

local function refresh_detail_title(title_buf, agent_id)
  local entry
  for _, a in ipairs(smelt.agent.list()) do
    if a.agent_id == agent_id then entry = a; break end
  end
  local line = " " .. agent_id
  local id_end = #line
  if entry then
    if entry.status == "idle" then
      line = line .. " \u{2713}"
    end
    if entry.task_slug and entry.task_slug ~= "" then
      line = line .. " \u{00b7} " .. entry.task_slug
    end
  end
  local snap = find_snapshot(smelt.agent.snapshots(), agent_id)
  if snap then
    if snap.context_tokens and snap.context_tokens > 0 then
      line = line .. "  " .. format_tokens(snap.context_tokens)
    end
    if snap.cost_usd and snap.cost_usd > 0 then
      line = line .. "  " .. format_cost(snap.cost_usd)
    end
  end
  smelt.api.buf.set_lines(title_buf, { line, "" })
  smelt.api.buf.add_highlight(title_buf, 1, 1, id_end, { fg = "agent", bold = true })
end

local function split_lines(s)
  local out = {}
  for line in (s .. "\n"):gmatch("(.-)\n") do
    table.insert(out, line)
  end
  while #out > 0 and out[#out] == "" do
    table.remove(out)
  end
  return out
end

local function refresh_detail_body(detail_buf, agent_id)
  local snap = find_snapshot(smelt.agent.snapshots(), agent_id)
  if not snap then
    smelt.api.buf.set_lines(detail_buf, { "(agent not tracked)" })
    return
  end

  local lines = { "Prompt:" }
  local dim_lines = { 1 }
  for _, raw in ipairs(split_lines(snap.prompt or "")) do
    table.insert(lines, " " .. raw)
  end
  table.insert(lines, "")
  if not snap.tool_calls or #snap.tool_calls == 0 then
    table.insert(lines, "(no tool calls yet)")
  else
    for _, entry in ipairs(snap.tool_calls) do
      local elapsed = ""
      if entry.elapsed_ms and entry.elapsed_ms >= 100 then
        elapsed = "  " .. format_duration(math.floor(entry.elapsed_ms / 1000))
      end
      table.insert(lines, string.format("%s %s%s", entry.tool_name, entry.summary, elapsed))
    end
  end
  smelt.api.buf.set_lines(detail_buf, lines)
  for _, i in ipairs(dim_lines) do
    local line = lines[i]
    if line then
      smelt.api.buf.add_dim(detail_buf, i, 0, #line)
    end
  end
end

local function open_detail(agent_id)
  local title_buf = smelt.api.buf.create()
  local detail_buf = smelt.api.buf.create()
  refresh_detail_title(title_buf, agent_id)
  refresh_detail_body(detail_buf, agent_id)

  local result = smelt.ui.dialog.open({
    panels = {
      { kind = "content", buf = title_buf, height = 2 },
      { kind = "content", buf = detail_buf, height = "fill",
        focusable = true, pad_left = 2 },
    },
    on_tick = function(ctx)
      refresh_detail_title(title_buf, agent_id)
      refresh_detail_body(detail_buf, agent_id)
    end,
  })
  return result
end

smelt.cmd.register("agents", function()
  smelt.spawn(function()
    while true do
      local agents = smelt.agent.list()
      if #agents == 0 then
        smelt.notify_error("no subagents running")
        return
      end

      local list_buf = smelt.api.buf.create()
      refresh_list(list_buf, agents, smelt.agent.snapshots())
      local title_buf = smelt.api.buf.create()
      smelt.api.buf.set_lines(title_buf, { "agents", "" })
      smelt.api.buf.add_dim(title_buf, 1, 0, #"agents")

      local result = smelt.ui.dialog.open({
        panels = {
          { kind = "content", buf = title_buf, height = 2 },
          { kind = "list", buf = list_buf, height = "fill" },
        },
        keymaps = {
          { key = "bs", hint = "\u{232b}: kill", on_press = function(ctx)
              local idx = ctx.selected_index
              if idx and agents[idx] then
                smelt.agent.kill(agents[idx].pid)
                agents = smelt.agent.list()
                refresh_list(list_buf, agents, smelt.agent.snapshots())
              end
            end },
        },
        on_tick = function(ctx)
          local fresh = smelt.agent.list()
          if agents_changed(fresh, agents) then
            agents = fresh
            refresh_list(list_buf, agents, smelt.agent.snapshots())
          end
        end,
      })

      if result.action == "dismiss" then return end
      local idx = result.option_index
      if not (idx and agents[idx]) then return end
      local detail = open_detail(agents[idx].agent_id)
      -- Detail dismissed → loop back and reopen the list view with
      -- refreshed data. Any other resolution falls out of the command.
      if detail.action ~= "dismiss" then return end
    end
  end)
end, { desc = "manage running agents" })
